persist: hook up an initial implementation of compaction

danhhz opened the pull request:
See the comments in trace.rs for an overview of how this is set up.

Compaction tests can quickly become a maintenance burden as it's
difficult to balance invariant verification against needing lots of
diffs when small compaction heuristics change. Invest early in
datadriven tests (the same style and underlying library as testdrive).

In production initially disabled by default, but turned on by the
`MZ_PERSIST_COMPACTION_ENABLED` env var for easy testing. In test,
enabled by default.
### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

I'd recommend taking a quick scan of the entire PR's diff to get a high level picture (maybe start with the comment in trace.rs) and then going back commit by commit.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

danhhz commented:
fyi @aljoscha @pH14

I had been hoping to get this out for review this week, but I'm coming up just shy on time (and focus). no need to review yet but absolutely do feel free to poke around and start leaving comments if you like

one big missing piece here that I owe is a justification for why I've forked Spine instead of using it directly. there are pros/cons to each and it'll take a bit to type up

I also have a couple final WIPs left and some testing to make sure that this doesn't mess up dev/CI even while disabled. I'd also like to make sure it doesn't noticeably regress the benchmarks, in particular open-loop

danhhz commented:
> some testing to make sure that this doesn't mess up dev/CI even while disabled

okay well that answers that. looks like some CI jobs (most notably testdrive) are _significantly_ slower with this PR in its current state, so I'll have to figure out what's going on there and a way to address it before this merges

danhhz commented:
my initial guess for the slowdown was that we're doing non-trivial work in decode now (ingesting batches 1:1 and letting the Spine decide where they end up), so put together a quick benchmark for state encode and decode. it's slower but I can't imagine this is enough to explain testdrive given the magnitude of the numbers. this would be a nice benchmark to merge, but it would require making quite a few things pub that currently are not, so I think it deserves it's own PR if we decide it's worth the work

i don't immediately have a second guess of what is causing the slowdown. i suppose my next steps are to read over the code again with a critical eye for that and/or running open loop (the latter I wanted to do anyway)

before (on my laptop)
```
MZ_PERSIST_RECORD_COUNT_SMALL=64 MZ_PERSIST_RECORD_SIZE_BYTES_SMALL=1024 MZ_PERSIST_BATCH_MAX_COUNT_SMALL=8
plumbing/encode_state/state/64KiB
                        time:   [1.6603 us 1.6613 us 1.6626 us]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

plumbing/decode_state/state/64KiB
                        time:   [3.6962 us 3.7005 us 3.7057 us]
Found 10 outliers among 100 measurements (10.00%)
  10 (10.00%) high severe
```

after (also on my laptop)
```
MZ_PERSIST_RECORD_COUNT_SMALL=64 MZ_PERSIST_RECORD_SIZE_BYTES_SMALL=1024 MZ_PERSIST_BATCH_MAX_COUNT_SMALL=8
plumbing/encode_state/state/64KiB
                        time:   [1.7780 us 1.7793 us 1.7807 us]
                        change: [+7.2663% +7.4660% +7.6840%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

plumbing/decode_state/state/64KiB
                        time:   [8.1047 us 8.1108 us 8.1183 us]
                        change: [+119.15% +119.45% +119.76%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  5 (5.00%) high severe
```

danhhz commented:
yup still repros. I'm a little stumped because I have a bunch of theories for things that might have gotten slower (added "empty batch optimization is not as comprehensive as before"), but so far nothing to explain why "as of ... not yet available". failed theories for asof are:
- bug in new upper calculation (should readily repro locally)
- slowness is causing starvation so the upper never gets updated (feels like more should be broken if it's bad enough to consistently break github-5983.td, which is as simple as it gets contention-wise)

about the only idea I have left atm is to hack together a change that speeds up decode by directly constructing spine's inner state. that should get it back to the same range it was before, but it's a long shot. I might instead take the rest of my day to circle back to the commentary I meant to write for this and let the testdrive problem mull in the background

danhhz commented:
As promised, here are the pros/cons that I have thought of for forking Spine vs using it directly:

Forking Spine
- (1) pro: implements `Send` and so works with persist's existing code structure. the dd Spine is not `Send` because of some internal `Rc`s and so would require us to switch to a model where we communicate with `Machine` over channels. in the long term, that may be a beneficial thing to do anyway (e.g. could help reduce CaS expectation failures when a reader+writer are in the same process), but I don't want to block compaction on it. so in the short term, this is ultimately why I forked in this PR
- (2) pro: allows us to mess directly with the internals of Spine. e.g. we could serialize/deserialize the spine internal levels directly (vs decode working by reintroducing all the batches and letting them advance though the spine logic as normal). if we have to speed up decode (as I've wondered in comments above) this is basically our only knob, so I'd like to keep it in the back pocket
- (3) pro: being able to litter in printlns and such during development was pretty invaluable to getting this all working. this is also possible if we used dd spine by pointing a Cargo.toml override at a local dd checkout, but it's certainly easier
- (4) con: a vague sense of dissatisfaction at the duplication
- (5) con: the work of keeping the fork's changes minimal
- (6) con: having to explain why we forked it when engineers are curious

Using DD Spine Directly
- (7) pro: if we could figure out a way to hook up `Cursor` (I have ideas), an actual Spine impl plus some tweaks to Snapshot+Listen gets us close to an `Arrangement` that's based by persist. the latency of accesses would be wildly different vs the normal in-mem Arrangement, so it's unclear if this is actually useful. it's certainly a very cool idea though
- (8) pro: a vague sense of satisfaction
- (9) pro: picking up Spine improvements for free (for reference: there have been 2 PRs in the last year, one to improve the generics that we've tossed in the fork and one with a minor alloc reuse)
- (10) con: it'd be tough to discover new merge requests. we'd have to do it via `map_batches` I think?

in summary: I think we'll need to periodically re-evaluate whether forking is the right choice, but I feel it is while we're in the rapid development phase. also, I'm reasonably certain that (modulo `Send`-ness) the `Trace` interface here (the only thing exposed to the rest of the crate) is possible to implement without the fork, which would make it easier to reverse this decision if the pro/con math changes (1 and 7 are the ones I'm most looking at)

danhhz commented:
environmentd metrics for the entire testdrive-0 run

before this PR
```
testdrive-materialized-1     | environmentd: 2022-06-23T19:09:50.943696Z  INFO mz_http_util: mz_persist_state_decode_count 96321
testdrive-materialized-1     | environmentd: 2022-06-23T19:09:50.943698Z  INFO mz_http_util: mz_persist_state_decode_seconds 3.5075530180000105
testdrive-materialized-1     | environmentd: 2022-06-23T19:09:50.943700Z  INFO mz_http_util: mz_persist_state_encode_count 193109
testdrive-materialized-1     | environmentd: 2022-06-23T19:09:50.943702Z  INFO mz_http_util: mz_persist_state_encode_seconds 3.2389715050000243
```

with this PR and compaction off
```
testdrive-materialized-1     | environmentd: 2022-06-23T20:17:01.711760Z  INFO mz_http_util: mz_persist_state_decode_count 124915
testdrive-materialized-1     | environmentd: 2022-06-23T20:17:01.711761Z  INFO mz_http_util: mz_persist_state_decode_seconds 892.4209364210026
testdrive-materialized-1     | environmentd: 2022-06-23T20:17:01.711763Z  INFO mz_http_util: mz_persist_state_encode_count 250230
testdrive-materialized-1     | environmentd: 2022-06-23T20:17:01.711765Z  INFO mz_http_util: mz_persist_state_encode_seconds 7.98944095699992
```

with this PR and compaction on (DNF because of the following, which seems to be a bug in this PR)
```
thread 'tokio:work-16' panicked at 'internal error: invalid encoded state: couldn't decode batch at key f81887b9-95f2-4e5a-9f16-b6e6294721cc: timestamp 1656011549309 is greater than or equal to the batch upper: Description { lower: Antichain { elements: [1656011524327] }, upper: Antichain { elements: [1656011544684] }, since: Antichain { elements: [0] } }', /Users/dan/mz/materialize/src/persist-client/src/read.rs:634:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I don't think we needed any more confirmation beyond what aljoscha has already found, but if we did this is pretty definitively. for slightly more state decode calls 124k vs 90k we spent 890s vs 3.5s in those calls

danhhz commented:
Okay, I think I may have figured out the `invalid encoded state: couldn't decode batch at key` issue. We were hardcoding a since of the minimum element in all the parts we wrote (correct for data coming in to persist, not correct for data coming out of compaction). The validation logic already correctly accounted for the case when since is beyond upper, but since the since was wrong, we were hitting the wrong branch. This 100% deserves a regression test in some way. I'm holding off on pushing the fix because I don't want CI to cancel my jobs :)

danhhz commented:
> almost certainly a correctness bug in compaction. I wonder if maelstrom could find it!

hell yes it does! local repro for the win!

```
INFO [2022-06-23 14:35:08,075] n0 stderr - maelstrom.process 2022-06-23T21:35:08.075088Z  INFO persistcli::maelstrom::node: res: {"src":"n0","dest":"c2","body":{"type":"error","msg_id":5499,"in_reply_to":994,"code":13,"text":"invalid read at time 360"}}
```

danhhz commented:
ran out of time for the day but I have most of a regression test for the "minimum timestamp since" bug. along the way, I found another bug in ColumnarRecords flushing (only affects tests) so it's 🐢 regression tests all the way down 🐢. I'll finish that up tomorrow and start looking into the "retractions for row that does not exist" bug.

@aljoscha @pH14 given the above results, I kinda think I should abandon the "merge compaction disabled by default" plan and figure out how to merge it enabled by default. i was mostly worried about the write amp, but it seems like on the balance that's a much lesser problem than state decoding. I was planning on circling back to do more correctness testing, but, given that we've already found 2, it seems worthwhile to pull that up and do it before this lands. a good start is probably the compaction tests that ruchir wrote for old persist, which I'd like to try datadriven-ifying. thoughts/opinions?

danhhz commented:
it seems there's still a correctness bug here, which I've been looking into this morning. testdrive-2 failed with

```
> SELECT * FROM v1;
--
  | rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 21s
  | gh6744.td:45:1: error: non-matching rows: expected:
  | [["176", "176", "14", "14"]]
  | got:
  | [["176", "176", "14", "14"], ["176", "176", "14", "14"], ["176", "176", "14", "14"], ["176", "176", "14", "14"], ["176", "176", "14", "14"]]
  | Poor diff:
  | + 176 176 14 14
  | + 176 176 14 14
  | + 176 176 14 14
  | + 176 176 14 14
```

it doesn't seem to be deterministic, testdrive-2 passed when run a second time at the same sha. it also didn't repro locally by repeatedly running `gh6744.td` in a bash loop. I wrote a bunch more tests (including porting anything I thought might be relevant from the old persist compaction tests) and discovered no new bugs

re-read the new `Compact::compact` impl closely and didn't see any obvious issues. the next things I'll try are to re-read `apply_merge_res` closely and also to think of some asserts/debugging info we could add to get more info on a repro

danhhz commented:
The best leads I have right now are:
- github-10587.td has reproed twice for me locally but it takes FOREVER to run. possible that runnings things in --release will help with that
- the "replica isolation" mzcompose seems like it repros more often than other things on CI. i haven't tried running it locally yet

the second repro on the testdrive file had my "updates are equal before and after compaction" assertion, but it didn't fire

danhhz commented:
tried a trimmed down version of github-10587.td with 1 source containing 1000 updates and got no repro. tried the same but the original 100000 updates and got a third repro. possibly we're on to something!

danhhz commented:
one thing I worry about here is that the github-10587.td failures look like (a bunch of data is all missing)

```
test/testdrive/github-10587.td:36:1: error: non-matching rows: expected:
[["100000"]]
got:
[["0"]]
Poor diff:
+ 0
- 100000
```

while in testdrive CI there are also some failures that look like (part of the data is duplicated)

```
kafka-avro-debezium-transaction.td:356:1: error: non-matching rows: expected:
--
  | [["(1,1)"], ["(2,3)"], ["(2,7)"], ["(3,9)"], ["(4,5)"], ["(8,9)"]]
  | got:
  | [["(1,1)"], ["(2,3)"], ["(2,7)"], ["(2,7)"], ["(3,9)"], ["(4,5)"], ["(8,9)"], ["(8,9)"]]
  | Poor diff:
  | + (2,7)
  | + (8,9)


```

not sure if that's 2 bugs or the same bug manifesting in 2 ways. i'm sure hoping the latter since this github-10587.td one seems easier to repro, but... I'm not hopeful, those are two pretty different manifestations

danhhz commented:
okay tried github-10587.td with 1000 updates (divided by 100) but PersistConfig::blob_target_size set to 1MB (divided by 128) to see if i could get a faster repro. nope

tried with MZ_PERSIST_COMPACTION_DISABLED=1 and 100000 updates and it _did_ repro. verified via eprintln logging that MZ_PERSIST_COMPACTION_DISABLED indeed is working, so looks like that bug at least is not with compaction. presumably it's with the changes needed to set up compaction. the snapshot on the failure had a batch that thought it had 100000 updates. oddly, the since of the other two batches is non-zero which doesn't match my mental model with compaction off. gonna investigate why noting came out of that len=100000 batch and also why these sinces are non-zero

danhhz commented:
okay one minor mystery solved! the non-zero sinces (even with compaction off) were coming from the optimization that just completes the merge if both side of a merge are empty

```
        // Special case empty batches.
        if self.b1.len() == 0 && self.b2.len() == 0 {
            return SpineBatch::Merged(HollowBatch {
                desc,
                keys: Vec::new(),
                len: 0,
            });
        }
```

as for the major mystery, i've now seen 100000 updates come out of SnapshotIter, so perhaps they're getting consolidated away after persist or something?

danhhz commented:
Confirmed that these updates are emitted at the same timestamp and with a diff of 1. After seeing a few instances of only one of the two snapshot parts in the logs before the testdrive error was emitted, accompanied by some instances of the following error, I am now suspicious that this is testdrive somehow timing out. I tried to read the testdrive code to confirm, but it's complicated enough that I stopped going down that particular rabbit hole. Recompiled everything with --release and haven't seen this fail since. Gonna ask around (aka @benesch) if anything in his mental model can explain what I'm seeing here before digging into this further.

```
environmentd: 2022-06-28T23:39:40.462822Z ERROR mz_environmentd::tcp_connection: error handling connection in pgwire server: connection closed
```

now moving to try reproing the replica-isolation failure on a bin/scratch machine

benesch commented:
I'm sharded too many ways at the moment so I didn't have time to grok all the context here. But I took a peek at a few of the red jobs in CI. I'm assuming those are the failures you're looking into? Did you check out services.log for each of the failed jobs? There are some very sketchy panics around:

```
replica-isolation-computed_1_1-1  | thread 'timely:recv-1' panicked at 'assertion failed: read > 0', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/tcp.rs:67:9
```

```
testdrive-materialized-1     | thread 'timely:work-0' panicked at 'failed to find index', /cargo/git/checkouts/differential-dataflow-d065d23d797aa027/5a013e0/src/operators/reduce.rs:974:59
```

danhhz commented:
yup! I'm hoping those are the same underlying cause as the testdrive failures, which seems(?) to be persist occasionally emitting duplicate data. all of which I'm still actively debugging. that ping was a testdrive question I have for you, happy to save it and ask in person since we happen to have a 1:1 tomorrow

benesch commented:
Hah, sorry, I'm happy to try to answer now—I just got confused about what your question was.

```
environmentd: 2022-06-28T23:39:40.462822Z ERROR mz_environmentd::tcp_connection: error handling connection in pgwire server: connection closed
```

This could be benign. We're way too aggressive about logging things that are perfectly graceful terminations as errors. It's quite possibly that testdrive triggers this log line whenever it shuts down, regardless of whether it timed out or not.

danhhz commented:
sure! basically, wondering if seeing this failure from the following testdrive file is consistent with the query simply timing out in some way (assume I was running locally in dev mode with some unmerged persist changes and lots of extra debugging so things are slower than usual). (also ignore the mismatched line number, I had trimmed down the file locally while iterating.)

```
test/testdrive/github-10587.td:36:1: error: non-matching rows: expected:
[["100000"]]
got:
[["0"]]
Poor diff:
+ 0
- 100000
```

https://github.com/MaterializeInc/materialize/blob/fca20c7397ab26cb53a1367e46049c8f8f3299ef/test/testdrive/github-10587.td#L129-L133

benesch commented:
Oh, if the query actually times out—i.e., literally never returns results—you should see "deadline exceeded" or some error like that. If testdrive is telling you the query returned 0, then it returned 0 (modulo some sort of massive bug in testdrive).

danhhz commented:
Ah hmmmm... just to check: that's true even if the sql query does return `[["0"]]` the first few times it's called, but then takes a lot longer to return on the Nth retry?

benesch commented:
Oh man, now you’re asking hard questions. I think you are correct that if it returns zero once and _then_ blocks forever on a retry that testdrive will tell you about the zero rather than the timeout. 😬

https://dev.materialize.com/api/rust/mz_ore/retry/struct.Retry.html#method.retry_async_canceling

> If f is forcibly canceled, the error returned will be the error returned by the prior invocation of f. If there is no prior invocation of f, then an Elapsed error is returned. The idea is that if f fails three times in a row with a useful error message, and then the fourth attempt is canceled because the timeout is reached, the caller would rather see the useful error message from the third attempt, rather than the “deadline exceeded” message from the fourth attempt.

danhhz commented:
okay cool, that matches my read of the code! (and explains why i saw it fail consistently in debug but not in release.) thanks! and sorry for subscribing you to this thread. feel free to unsubscribe! :D

danhhz commented:
> @aljoscha curious if you have any good ideas tomorrow for how to debug this

probably nevermind, I _think_ the replica-isolation test has a good repro when i run it on a bin/scratch machine

danhhz commented:
okay just got a repro on `source2` with only the "drop-create-replica" and "create-invalid-replica" disruptions. As usual it was in the second disruption "create-invalid-replica". `source2` is a new source created after the disruption, so it really feels like i should be able to repro this with just a large amount of data. still haven't managed to do that yet though

danhhz commented:
meanwhile, https://github.com/MaterializeInc/materialize/pull/13160, which is just all the commits of this up but not including https://github.com/MaterializeInc/materialize/pull/13160/commits/c4cc094198847656ff8eb1d414e36e40cd91dc40 has passed replica-isolation in CI 3 times now. is that enough to consider it working? if so, i could try bisecting in bits of c4cc094198847656ff8eb1d414e36e40cd91dc40. I've read that over several times now and don't really see what in it could be causing this issue.

danhhz commented:
I'm starting to feel like i'm losing my mind. one of the failures in CI on #13358 (this plus even more debugging printlns) was e.g. kafka_sources/upsert-modification-{before,after}.td in the persistence mzcompose. and it's a small source! just a few rows! no idea why it wasn't failing on this PR, but i'll take it

```
kafka-sources/upsert-modification-after.td:43:1: error: non-matching rows: expected:
[["bírdmore", "géese", "0", "6"], ["fish", "fish", "0", "1"], ["mammal1", "mouse", "0", "9"], ["mammalmore", "moose", "0", "7"]]
got:
[["bírdmore", "geese", "0", "3"], ["bírdmore", "géese", "0", "6"], ["fish", "fish", "0", "1"], ["fish", "fish", "0", "1"], ["mammal1", "moose", "0", "4"], ["mammal1", "mouse", "0", "9"], ["mammalmore", "moose", "0", "7"]]
Poor diff:
+ bírdmore geese 0 3
+ fish fish 0 1
+ mammal1 moose 0 4
```

```
./bin/mzcompose --find persistence down -v && MZ_PERSIST_COMPACTION_DISABLED=1 TD_TEST=upsert-modification ./bin/mzcompose --find persistence run kafka-sources
```

i ran it locally using the above command and got a repro. then i deleted the unrelated sources in the upsert-modification files and.. it no longer reproed? then i reverted and it reproed again. I'm starting to wonder if this is somehow timing based. like maybe we need the bigger kafka source to be there. i can't think of anything in persist that would be timing based though

the test in question writes a kafka topic with the following:

```
[1] add fish
[2] add bìrd1
[3] add bírdmore
[4] add mammal1
[5] delete bìrd1
<restart>
[6] overwite bírdmore
[7] add mammalmore
[8] delete mammal1
[9] add/overwrite mammal1
```
so at the end of the test, we're supposed to have [1][6][7][9]. we had those _plus_ [1][3][4] (including a dup of [1]). both times this failed on CI we had _this exact same extra data_. that's kinda wild and i have no mental model for why those rows. note that my local repro has compaction disabled. (replica-isolation, when it fails with extra rows, seems to get a different number of extra rows each time)

danhhz commented:
GOT IT!

I'm completely wiped, so a description of the bug and how it explains various things that I saw will have to wait for tomorrow. As it often is with stuff like this, it's real dumb and actually even a concern I thought of a couple of times in the past and then just forgot about :facepalm: . I have some ideas for a regression test and maybe a tweak to maelstrom so it would have also been caught there.

philip-stoev commented:
@danhhz I was rooting for you as I read through the github emails! Please open tickets for any enhancements/changes to testdrive, mzcompose and/or the isolation or persistence tests or any other workflows that you would like to see as a result of your debugging odyssey and assign to me.

danhhz commented:
thanks @philip-stoev! I actually considered this last night and didn't come up with a ton. the overwhelming takeaway was that we wouldn't have caught this bug so early without the tests! (probably would have had a 1000x harder time debugging it when some customer hit it, and hopefully that happened before GA.) here's the notes I do have:
- keep doing: the bug indeed ended up being timing related. it seemed important to have some amount of load on the system to get the bad timing. concretely: having both a big source along with a small one in the upsert-modification test in test/persistence (which is the test that ultimately gave me the diagnosis) was a blessing
- maybe tweak: one thing I noticed is maybe the different "disruptions" in the replica-isolation test were leaking between each other. I'm definitely not sure, but one bit I didn't know offhand was what was being used for the persist consensus impl and the catalog stash and how that was being reset in between disruption cases. in the past, the catalog would have been sqlite and covered when the volumes were wiped. unclear if maybe the transition to postgres means this is no longer happening?

i'd also be interested in running this PR through the nightlies. i saw you post somewhere in slack recently how to do that. if I do, any guidance on figuring out what, if anything, is a real failure vs a flake?

danhhz commented on src/persist-client/src/read.rs:
```diff
@@ -624,31 +673,26 @@ where
         drop(value);
 
-        let mut ret = Vec::new();
+        // WIP this is really subtle and possibly wrong, justify!
```
@aljoscha I was talking to paul yesterday about this. we currently have two "kinds" of batches in persist. one is a batch written directly by a client, which might be registered with truncated bounds and will always have a since of the minimum timestamp. the other is a batch written by compaction, which always is registered with bounds that exactly match the data in blob and the since might be downgraded. we need to perform this filtering logic on the first but not the second (and in the general case, couldn't perform it because the since might have erased the distinction in the original timestamps).

i'm not sure what to do about this/how to structure this code in a satisfying way, but it feels a bit funky right now, so calling it out and I'd love your thoughts

pH14 replied:
Not familiar enough to know if the current `since.less_equal(lower)` behavior is expected to be a guarantee going forward, but agreed that it feels subtle... feels like we want one more bit of information

danhhz replied:
wrote a much bigger comment here and restructured the code a bit. i still don't love it, so happy to accept any suggestions if you have them!

aljoscha replied:
I think with the latest change (big comment) this is fine to merge/I didn't come up with anything better after pondering it for a while. Though I agree that it's funky!

danhhz commented on src/persist-client/src/impl/trace.rs:
```diff
@@ -0,0 +1,1074 @@
+            }
+            match batch {
+                MergeState::Double(MergeVariant::InProgress(batch1, batch2, m)) => {
+                    // WIP make the two batch clones share internals
```
thoughts? this is something we'd need to do anyway to reuse the real Spine. I'm thinking something like this (with better names)

```
struct TraceBatch<T> {
   desc: Description<T>
   len: usize,
   core: Arc<Mutex<TraceBatchCore<T>>>,
}
enum TraceBatchCore<T> {
    Merged(HollowBatch<T>)
    Fueled(Vec<HollowBatch<T>>)
}
```

to make things work with real spine, we could register an Arc of the "core" of each batch as it's inserted into the spine. then apply_merge_res could mutate it via the one in the registry. this structure has the advantage that the desc and len never change out from under Spine. this is not true for len as things are typed right now, but it seems like a good property to have. after swapping a merge res in, the len of the TraceBatch and the one of the core might disagree, but this is fine since the former is an upper bound on the latter. we'd have to figure out how to remove things from the registry once spine is finished with them, maybe something with the spine logger or using a weak Arc in the registry?

danhhz replied:
indeed we do have to update the ones in the FuelingMerge because when that merge finishes, those are the ones that get turned into the new batch (this was a bug I had to track down at some point before this was even a PR 😅 )

https://github.com/MaterializeInc/materialize/pull/13160/files#diff-d74ca1805afb882373ab284ccd11e8eedda43152b4f31d7f309941a3b894d3c9R293-R334

the code we had before was fine without the Arc+SpineBatchCore, we only need this new thing if we want to work with regular Spine. i'm pretty divided on which of the two we should start with. thoughts?

aljoscha replied:
The diff in Spine that we need when we don't have `Arc+SpineBatchCore` is that we need to patch things up "outside" the batches, right? I think I like that slightly more, because it doesn't rely on the somewhat hidden shared nature of `SpineBatch` for correctness. But I'd also like it if we stayed true to vanilla `Spine` as much as possible. Though call!

pH14 commented on src/persist-client/tests/trace/since_upper:
```diff
@@ -0,0 +1,54 @@
+
+# A [0,1) batch advances the upper to [1]
+insert
+0 1 0 0
```
fwiw, I found most of the datadriven tests pretty readable without digging into their syntax, but this one tripped me up

danhhz replied:
good to know! i can totally see it. any thoughts on what to do instead?

pH14 reviewed (commented):
I did a fairly superficial read through of all of this (and might have avoided the Trace/Spine implementation for the first pass), and to this layperson, I think it makes sense! I like how the compaction work slots in nicely with the existing state machine, and shouldn't be too hard to move around if we need to going forward.

Oh and do we have test coverage around the `next_listen_batch` bug you/Aljoscha found?

danhhz commented on src/persist-client/src/impl/state.rs:
```diff
@@ -515,6 +491,68 @@ mod codec_impls {
+                        x
+                    )
+                }
+                // TODO: Can we rehydrate the Spine more directly?
```
yep! though I think the invalidations are a problem and one fixable in other ways that I'd like to explore first. rewrote this comment a bit, let me know what you think

danhhz commented on src/persist-client/src/impl/state.rs:
```diff
@@ -30,13 +31,19 @@ pub struct ReadCapability<T> {
     pub since: Antichain<T>,
 }
 
+#[derive(Clone, Debug, PartialEq)]
```
done!

danhhz commented on src/persist-client/src/impl/trace.rs:
```diff
@@ -0,0 +1,1044 @@
+                SpineBatch::Merged(_) => return false,
+                SpineBatch::Fueled { desc, parts } => (desc, parts),
+            };
+            // We could be more lenient here and swap out a subset if the descs
```
expanded it a bit. any clearer now?

danhhz commented on src/persist-client/src/impl/compact.rs:
```diff
@@ -0,0 +1,294 @@
+        let mut should_compact =
+            req.inputs.len() >= 16 || req.inputs.iter().map(|x| x.len).sum::<usize>() >= 128;
+        // TODO: Enable the heuristics after we've gotten a sense for where
+        // we're starting.
```
my convention (just so you know for the future) is that WIP generally means I intended to do it before merging and TODO means not. though, funny enough, I've started to warm up to the idea of enabling them from the beginning. I don't see any real reason not to and it might help avoid giving a bad initial impression of persist compaction to the rest of the company

metrics are a good idea! done!

danhhz commented on src/persist-client/src/impl/compact.rs:
```diff
@@ -0,0 +1,294 @@
+            let mut updates = Vec::new();
+            for part in req.inputs.iter() {
+                for key in part.keys.iter() {
+                    handle.block_on(fetch_batch_part(
```
it gets a lot harder once we are doing an N way merge with bounded memory. it's a good idea, but i'd like to avoid adding complexity here before we do the bounded memory bit so we don't get in our own way :)

danhhz reviewed (commented):
ended up pushing a big batch of polish, WIP resolution, tests, and review comment responses in one commit. i'll break these last two WIP commits back up into the original commit structure at some point, but it's gonna take some time and this plays more nicely with github reviews anyway

the impl of the datadriven tests is a hot mess at the moment. i'm still mulling over how to clean that up, but tbh i'm not sure it needs to block merging this PR. curious if either if you feel differently?

> This span should be created like this

ah good point. done! (aka merged in the commit you pushed to my branch)

> Oh and do we have test coverage around the next_listen_batch bug you/Aljoscha found?

yup! added in this last round of updates. I'd like to add maelstrom coverage too, but I ran out of time for the day

aljoscha reviewed (commented):
Pushing out first round of questions/comments.

aljoscha commented on src/persist-client/tests/trace/empty_batch_optimization:
```diff
@@ -0,0 +1,54 @@
+[0][3][0] 0
+
+# Now insert a non-empty batch. This happens to roll up into the empty one and
+# generate a merge.
```
nit: the fact that this generates a merge is not visible in the test.

aljoscha replied:
that's what I was suspecting. 😅

aljoscha commented on src/persist-client/src/batch.rs:
```diff
@@ -422,25 +459,50 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
     }
 }
 
+pub(crate) fn truncate_batch<T: Timestamp>(
```
nit: could be called `verify_truncate_batch()` or `check_truncate_batch()` or `is_truncate_batch()`, because the current name suggests to me that it does the truncation.

aljoscha replied:
👌

aljoscha reviewed (approved):
I've now also read and understood the datadriven tests! 🙌

I had a couple nits/questions but I think the overall approach works well and you probably got out most (all?) bugs by now. We should merge now, to give this as much exposure as we can, so that we can shake out bugs/performance issues.

philip-stoev commented:
I will now schedule a Nightly run. To run a Nightly yourself, go to the Nightly page in Buildkite, click New Build . Put the name of your branch in the 1st and 3rd box and the full git sha of your commit in the middle box. Then, once the build starts, you will be given an option to run all or only some of the individual jobs. Figuring out if there is a regression or not is unfortunately a matter of checking if the failures of Nightly against "main" are identical or different from the failures in your branch.

danhhz reviewed (commented):
> you probably got out most (all?) bugs by now

😅 🙈

danhhz commented:
okay rebased! everything since the last time paul looked at this is in the fixup commit. the commit structure isn't perfect -- a few of the things in the fixup commit should get amended into something other than "persist: hook up an initial implementation of compaction" -- but this was a tough rebase and I don't want to look at it too hard and spook it. I didn't get to any tuning of the heuristic, but I might see how far I can get in an hour or two on Tuesday. putting this down over the weekend

philip-stoev reviewed (approved):
There are no branch-specific failures in Nightly. Performance as per the feature-benchmark is not worse than "main".

danhhz commented:
okay, did an extremely quick and dirty compaction heuristic tuning, erring on the side of high write amp since whatever we do here is strictly a write amp improvement over having the heuristic completely disabled. I suspect we'll have to tune this down further, but it's time to start getting some feel for compaction in the wild

pH14 commented:
🙇 for splitting out the recent commits, much easier to focus in on the different logical changes

`remove_redundant_merge_reqs` reads great to me! Not super concerned with the O(n^2) behavior given that it's only called on one insert at a time, seems like something we can measure and revisit in a future PR if benchmarking/real-life shows it being problematic

The compaction heuristic also LGTM and seems like a reasonable starting point

danhhz commented:
TFTRs!!

> 🙇 for splitting out the recent commits, much easier to focus in on the different logical changes

sorry it hasn't been that way all along! that's usually what I do. just ended up being hard here with all the debugging
